### PR TITLE
Xpetra: ETI VectorFactory

### PR DESCRIPTION
**Pull request by william76**
@trilinos/xpetra 

ETI conversion for VectorFactory

Relates to #4986

### CODE REVIEW

File: packages/xpetra/src/Vector/Xpetra_Vector.hpp [hunk 1]
```diff
@@ -61,10 +61,10 @@ namespace Xpetra {
     : public virtual MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node >
   {
   public:
-    typedef Scalar scalar_type;
-    typedef LocalOrdinal local_ordinal_type;
+    typedef Scalar        scalar_type;
```
Comment (mhoemmen) [inline packages/xpetra/src/Vector/Xpetra_Vector.hpp]: If you're going to change the spacing, why not also change these typedefs to C++11 - style `using` statements?

File: packages/xpetra/src/Vector/Xpetra_VectorFactory_decl.hpp [hunk 2]
```diff
@@ -248,15 +257,16 @@ namespace Xpetra {
...
+    typedef long long  GlobalOrdinal;
     typedef EpetraNode Node;
 
 #undef XPETRA_VECTORFACTORY_SHORT
 #include "Xpetra_UseShortNames.hpp"
 
   private:
+
     //! Private constructor. This is a static class.
     VectorFactory() {}
```
Comment (mhoemmen) [inline packages/xpetra/src/Vector/Xpetra_VectorFactory_decl.hpp]: Prefer `VectorFactory() = default;` to this approach.

File: packages/xpetra/src/Vector/Xpetra_Vector.hpp (hunk 1 above)
Comment (william76) [reply packages/xpetra/src/Vector/Xpetra_Vector.hpp]: heh, ok. My PR failed anyhow... i'm updating my local build to see if I can catch these for future testing.  I'll make these modifications with the next update.

**Comment by william76** [comment]
@jhux2 @csiefer2 I think I can use a fresh pair of eyes on this... I'm getting errors now that I've updated my build a bit to better mirror what PR testing is doing.  I don't see what I'm missing here, do you guys see anything jumping out that I've missed?

Here's the basic error message I'm seeing
```
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Vector/Xpetra_VectorFactory_decl.hpp:88:25: error: ‘Xpetra::VectorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Vector’ is not a template
     static Teuchos::RCP<Vector<Scalar,LocalOrdinal,GlobalOrdinal,Node>>
                         ^~~~~~
```
<details><summary>Expand for full error message</summary>
<pre>[ 41%] Building CXX object packages/xpetra/src/CMakeFiles/xpetra.dir/Utils/ExplicitInstantiation/Xpetra_VectorFactory.cpp.o
In file included from /ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Vector/Xpetra_VectorFactory_def.hpp:49:0,
                 from /ascldap/users/wcmclen/dev/trilinos-dev/build/xpetra-eti-dev/packages/xpetra/src/Utils/ExplicitInstantiation/Xpetra_VectorFactory.cpp:55:
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Vector/Xpetra_VectorFactory_decl.hpp:88:25: error: ‘Xpetra::VectorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Vector’ is not a template
     static Teuchos::RCP<Vector<Scalar,LocalOrdinal,GlobalOrdinal,Node>>
                         ^~~~~~
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Vector/Xpetra_VectorFactory_decl.hpp:89:36: error: ‘Xpetra::VectorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Map’ is not a template
     Build(const Teuchos::RCP<const Map<LocalOrdinal,GlobalOrdinal,Node>> &map, bool zeroOut=true);
                                    ^~~
In file included from /ascldap/users/wcmclen/dev/trilinos-dev/build/xpetra-eti-dev/packages/xpetra/src/Utils/ExplicitInstantiation/Xpetra_VectorFactory.cpp:55:0:
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Vector/Xpetra_VectorFactory_def.hpp:60:32: error: ‘Xpetra::VectorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Map’ is not a template
 Build(const Teuchos::RCP<const Map<LocalOrdinal,GlobalOrdinal,Node>>& map,
                                ^~~
make[2]: *** [packages/xpetra/src/CMakeFiles/xpetra.dir/Utils/ExplicitInstantiation/Xpetra_VectorFactory.cpp.o] Error 1
make[1]: *** [packages/xpetra/src/CMakeFiles/xpetra.dir/all] Error 2
make: *** [all] Error 2
</pre></details>

The code from my `Xpetra_VectorFactory_decl.hpp` and `Xpetra_VectorFactory_def.hpp` files is here:

https://github.com/trilinos/Trilinos/blob/88d0b69c7200a79852867c01de434332fcefcf56/packages/xpetra/src/Vector/Xpetra_VectorFactory_decl.hpp#L86-L90

https://github.com/trilinos/Trilinos/blob/88d0b69c7200a79852867c01de434332fcefcf56/packages/xpetra/src/Vector/Xpetra_VectorFactory_def.hpp#L56-L62

When I compare this to what I've done for other *Factory classes, I'm not seeing what I'm doing different here.  For example, the `Xpetra_MultiVectorFactory_decl.hpp` Build methods look like this:

https://github.com/trilinos/Trilinos/blob/88d0b69c7200a79852867c01de434332fcefcf56/packages/xpetra/src/MultiVector/Xpetra_MultiVectorFactory_decl.hpp#L79-L83

which looks to me like I'm doing the same stuff in VectorFactory... so why doesn't C++ think it's a type?  I've tried making these into `Xpetra::Vector<Scalar,LocalOrdinal,GlobalOrdinal,Node>` and I can get past compilation steps but then I get undefined reference linker errors like this:
```
undefined reference to 
`Xpetra::VectorFactory<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::
Build(Teuchos::RCP<Xpetra::Map<int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, bool)'
```
<details>
<summary>Expand for full error message</summary>
<pre>
[ 42%] Linking CXX executable Xpetra_MultiVector_UnitTestsThyraSpecific.exe
../../src/libxpetra.a(Xpetra_MapExtractor.cpp.o): In function `Xpetra::MapExtractor<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::getVector(unsigned long, bool, bool) const':
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/BlockedCrsMatrix/Xpetra_MapExtractor_def.hpp:594: undefined reference to `Xpetra::VectorFactory<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Build(Teuchos::RCP<Xpetra::Map<int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, bool)'
../../src/libxpetra.a(Xpetra_MapExtractor.cpp.o): In function `Xpetra::MapExtractor<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::ExtractVector(Teuchos::RCP<Xpetra::Vector<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const>&, unsigned long, bool) const':
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/BlockedCrsMatrix/Xpetra_MapExtractor_def.hpp:172: undefined reference to `Xpetra::VectorFactory<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Build(Teuchos::RCP<Xpetra::Map<int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, bool)'
../../src/libxpetra.a(Xpetra_MapExtractor.cpp.o): In function `Xpetra::MapExtractor<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::ExtractVector(Teuchos::RCP<Xpetra::Vector<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > >&, unsigned long, bool) const':
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/BlockedCrsMatrix/Xpetra_MapExtractor_def.hpp:200: undefined reference to `Xpetra::VectorFactory<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Build(Teuchos::RCP<Xpetra::Map<int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, bool)'
collect2: error: ld returned 1 exit status
make[2]: *** [packages/xpetra/test/MultiVector/Xpetra_MultiVector_UnitTestsThyraSpecific.exe] Error 1
make[1]: *** [packages/xpetra/test/MultiVector/CMakeFiles/Xpetra_MultiVector_UnitTestsThyraSpecific.dir/all] Error 2
make: *** [all] Error 2
</pre>
</details>

My nm log is this: [__output.nm.log](https://github.com/trilinos/Trilinos/files/3838040/__output.nm.log).

**Comment by william76** [comment]
@jhux2 @csiefer2 thinking on this a bit more... I looked again and I see that for other Factory methods like MapFactory we had to create the `Xpetra_EpetraMapFactory.cpp` files... I'll see if I can create a `Xpetra_EpetraVectorFactory.cpp` and see if that fixes the issue.

**Comment by jhux2** [comment]
@william76 Try adding `Xpetra::` namespace?

**Comment by william76** [comment]
@jhux2 I'm testing out a `Xpetra_EpetraVectorFactory.cpp` file to see if that's what was causing my woes... I'll update that wasn't the fix and things still woof.

**Comment by william76** [comment]
Hmmm, still failing but it looks like it's wanting a Scalar=long long, LocalOrdinal=int, GlobalOrdinal=long long ... that combination isn't among the list of partial specializations... @csiefer2 should I just create one for that or would it not make sense?  If I need to add it, which of the EPETRA guard sections would it need to be in?  Testing isn't finished but you can see the errors here: https://testing.sandia.gov/cdash/viewBuildError.php?buildid=5893352

**Comment by william76** [comment]
@csiefer2 @jhux2 I might be stuck again on this... The errors I get look like this:
```
../../src/libxpetra.a(Xpetra_MapExtractor.cpp.o): In function `Xpetra::MapExtractor<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::getVector(unsigned long, bool, bool) const':
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/BlockedCrsMatrix/Xpetra_MapExtractor_def.hpp:594: undefined reference to `Xpetra::VectorFactory<long long, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Build(Teuchos::RCP<Xpetra::Map<int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, bool)'
```

so it looks like it wants a Scalar=long long, LO=int, GO=long long variation... which doesn't exist in the specializations. Why it doesn't just use the fully templated version is confusing to me since isn't that what templates are for?  So, I tried just adding a specialization for a `VectorFactory<long long, int, long long, Node>` in my decl file and the EpetraVectorFactory.cpp file but that then fails for some other compiler issue:

```
[ 37%] Building CXX object packages/xpetra/src/CMakeFiles/xpetra.dir/Vector/Xpetra_EpetraVectorFactory.cpp.o
In file included from /ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/teuchos/core/src/Teuchos_ArrayRCPDecl.hpp:47:0,
                 from /ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/teuchos/core/src/Teuchos_ArrayRCP.hpp:46,
                 from /ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/teuchos/core/src/Teuchos_Array.hpp:52,
                 from /ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/tpetra/core/src/Tpetra_ConfigDefs.hpp:86,
                 from /ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Headers/Xpetra_ConfigDefs.hpp:90,
                 from /ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Vector/Xpetra_VectorFactory_decl.hpp:49,
                 from /ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Vector/Xpetra_EpetraVectorFactory.cpp:46:
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/teuchos/core/src/Teuchos_RCP.hpp: In instantiation of ‘Teuchos::RCP<T>::RCP(const Teuchos::RCP<T2>&) [with T2 = Xpetra::EpetraIntVectorT<long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >; T = Xpetra::Vector<long long int, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >]’:
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Vector/Xpetra_EpetraVectorFactory.cpp:247:75:   required from here
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/teuchos/core/src/Teuchos_RCP.hpp:298:38: error: cannot convert ‘Xpetra::EpetraIntVectorT<long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >*’ to ‘Xpetra::Vector<long long int, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >*’ in initialization
     node_(r_ptr.access_private_node())
                                      ^
```

I'm inclined to think that I should not have to add a new specialization since if it was necessary I'd expect that it would have been made already... but I thought it was worth a college try.  

So at this point, I'm lost on this one. Definitely need some additional eyes on it to see if we can figure out what's missing.

**Pull request merged**